### PR TITLE
[TF FE] Implement conversion for Attention OCR model

### PR DESCRIPTION
**Pull request by rkazants**
**Details:** The following scope of work is done to make Attention OCR convertable:
1. Refactored translators for BiasAdd, Slice, and ArgMax operations. Add translation for StopGradient operation.
2. The previous traversing algorithm to compute topological sorted nodes list was incorrect. Now it is implemented based on topologically_sorted function from core/graph_util.hpp.
3. The unsupported data types are now preliminary converted to undefined type for the purpose of to have them cut off.

**Tickets:** 88976, 88975

Signed-off-by: Kazantsev, Roman <roman.kazantsev@intel.com>

### CODE REVIEW

File: src/frontends/tensorflow/src/op/slice.cpp [hunk 1]
```diff
@@ -18,11 +18,22 @@ OutputVector translate_slice_op(const NodeContext& node) {
...
+    auto stop_neg = make_shared<Add>(make_shared<Add>(make_shared<ConvertLike>(input_shape, size), size), one);
+
+    // select the correct stop value based on a sign of size value
+    auto zeros = make_shared<Constant>(size.get_element_type(), Shape{}, 0);
+    auto negative_sizes_mask = make_shared<Less>(size, zeros);
+    auto stop = make_shared<Select>(negative_sizes_mask, stop_neg, stop_pos);
+
+    // broadcast step value
+    auto start_shape = make_shared<ShapeOf>(start);
+    auto step = make_shared<Broadcast>(one, start_shape);
```
Comment (cavusmustafa) [inline src/frontends/tensorflow/src/op/slice.cpp:37]: How this would behave if the first dimension of the input is 0? For example: (0, X, Y).

Comment (rkazants) [reply src/frontends/tensorflow/src/op/slice.cpp:37]: `start_shape` is 1D tensor.
In case input to be empty tensor, it will compute `stop` to be equal to `start` by this dimension and the output will be also empty tensor.

File: tools/mo/openvino/tools/mo/main.py [hunk 2]
```diff
@@ -373,13 +373,6 @@ def check_fallback(argv: argparse.Namespace):
 
 
 def prepare_ir(argv: argparse.Namespace):
```
Comment (itikhono) [inline tools/mo/openvino/tools/mo/main.py:375]: why is this deleted?

File: src/frontends/tensorflow/src/decoder_proto.cpp [hunk 3]
```diff
@@ -67,8 +67,14 @@ ov::Any DecoderProto::get_attribute(const std::string& name) const {
...
     }
 
-    case ::tensorflow::AttrValue::ValueCase::kType:
-        return TYPE_MAP().at(attrs[0].type());
+    case ::tensorflow::AttrValue::ValueCase::kType: {
+        if (TYPE_MAP().count(attrs[0].type())) {
+            return TYPE_MAP().at(attrs[0].type());
+        } else {
+            // for all unsupported types return undefined type
+            return ov::element::undefined;
```
Comment (itikhono) [inline src/frontends/tensorflow/src/decoder_proto.cpp:75]: should we throw exception here?

File: src/frontends/tensorflow/src/input_model.cpp [hunk 4]
```diff
@@ -170,82 +170,97 @@ std::vector<std::shared_ptr<OpPlace>> InputModel::InputModelTFImpl::get_op_place
...
+
+                // is_input is a flag to leave producer operation node or not.
+                // this producing node is not left if consumer is pruned by its input port,
+                // the producer node is pruned by its output port or the producer becomes new input
+                // 1. check if the current node is pruned by its input port
+                bool is_input = false;
+                std::string input_port_name = std::to_string(input_port_idx) + ":" + current_operation_name;
+                if (m_tensor_places.find(input_port_name) != m_tensor_places.end()) {
+                    const auto& tensor_place = m_tensor_places[input_port_name];
+                    is_input = is_input || (tensor_place->is_input() ? true : false);
```
Comment (itikhono) [inline src/frontends/tensorflow/src/input_model.cpp:228]: is_input() returns bool
```suggestion
                    is_input = is_input || tensor_place->is_input();
```

Comment (mvafin) [inline src/frontends/tensorflow/src/input_model.cpp:228]: Can it be simplified like this:
`is_input = is_input || tensor_place->is_input();`

File: src/frontends/tensorflow/src/input_model.cpp [hunk 5]
```diff
@@ -170,82 +170,97 @@ std::vector<std::shared_ptr<OpPlace>> InputModel::InputModelTFImpl::get_op_place
...
+                }
+
+                // 2. check if the producer node is pruned by its output port
+                std::string output_port_name = producer_name + ":" + std::to_string(producer_output_port_idx);
+                if (m_tensor_places.find(output_port_name) != m_tensor_places.end()) {
+                    const auto& tensor_place = m_tensor_places[output_port_name];
+                    is_input = is_input || (tensor_place->is_input() ? true : false);
```
Comment (mvafin) [inline src/frontends/tensorflow/src/input_model.cpp:235]: Same comment:
`is_input = is_input || tensor_place->is_input();`

File: src/frontends/tensorflow/src/input_model.cpp [hunk 6]
```diff
@@ -170,82 +170,97 @@ std::vector<std::shared_ptr<OpPlace>> InputModel::InputModelTFImpl::get_op_place
...
+                }
+
+                // 3. check if the current node is an input
+                FRONT_END_GENERAL_CHECK(m_op_places_map.count(producer_name),
+                                        "There is no operation node with name: " + producer_name);
+                const auto& producer_operation_place = m_op_places_map.at(producer_name);
+                if (m_tensor_places.find(producer_name) != m_tensor_places.end()) {
+                    const auto& tensor_place = m_tensor_places[producer_name];
+                    is_input |= (tensor_place->is_input() ? true : false);
```
Comment (mvafin) [inline src/frontends/tensorflow/src/input_model.cpp:244]: Again:
`is_input |= tensor_place->is_input();`

File: src/frontends/tensorflow/src/op/slice.cpp [hunk 7]
```diff
@@ -18,11 +18,22 @@ OutputVector translate_slice_op(const NodeContext& node) {
...
+    // compute stop values in case non-negative sizes
+    auto stop_pos = make_shared<Add>(start, size);
 
-    auto one = make_shared<Constant>(element::i64, Shape{1}, 1);
-    auto shape = make_shared<ShapeOf>(start);
-    auto step = make_shared<Broadcast>(one, shape);
+    // compute stop values in case positive sizes
+    auto input_shape = make_shared<ShapeOf>(input);
+    auto one = make_shared<Constant>(size.get_element_type(), Shape{}, 1);
+    auto stop_neg = make_shared<Add>(make_shared<Add>(make_shared<ConvertLike>(input_shape, size), size), one);
```
[the remark this answers is not in the document]
Comment (rkazants) [reply src/frontends/tensorflow/src/op/slice.cpp:27]: `ConvertLike` is in our opset, please check it :)
I will think of the optimization for this chunk

File: src/frontends/tensorflow/src/decoder_proto.cpp (hunk 3 above)
Comment (itikhono) [reply src/frontends/tensorflow/src/decoder_proto.cpp:75]: discussed offline.
this is used for 1st convertion before cutting

File: src/frontends/tensorflow/src/op/slice.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (itikhono) [reply src/frontends/tensorflow/src/op/slice.cpp:27]: I believe ConvertLike is in our opset.

File: src/frontends/tensorflow/src/op/slice.cpp [hunk 8]
```diff
@@ -18,11 +18,22 @@ OutputVector translate_slice_op(const NodeContext& node) {
...
     auto size = node.get_input(2);
 
-    auto stop = make_shared<Add>(start, size);
+    // compute stop values in case non-negative sizes
+    auto stop_pos = make_shared<Add>(start, size);
 
-    auto one = make_shared<Constant>(element::i64, Shape{1}, 1);
-    auto shape = make_shared<ShapeOf>(start);
-    auto step = make_shared<Broadcast>(one, shape);
+    // compute stop values in case positive sizes
```
Comment (itikhono) [inline src/frontends/tensorflow/src/op/slice.cpp:24]: "in negative sizes" you mean?

File: src/frontends/tensorflow/src/op/slice.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (itikhono) [reply src/frontends/tensorflow/src/op/slice.cpp:27]: 1d tensor = [1, 2, 3, 4]
dim = 4, let's say size = -2
size -2 points to [1, 2, **3**, 4] element
your formula: ind = dim - size + 1? 
ind = 4 - 2 + 1 = 3 points to [1, 2, 3, **4**] 

I hope our Slice equals python operations
image

[the remark this answers is not in the document]
Comment (itikhono) [reply src/frontends/tensorflow/src/op/slice.cpp:27]: it looks like tf.slice doesn't support negative sizes
image

Comment (itikhono) [inline src/frontends/tensorflow/src/op/slice.cpp:27]: it looks like tf.slice doesn't support negative sizes
image

Comment (itikhono) [reply src/frontends/tensorflow/src/op/slice.cpp:27]: except -1 :)

File: tools/mo/openvino/tools/mo/main.py (hunk 2 above)
Comment (rkazants) [reply tools/mo/openvino/tools/mo/main.py:375]: Will be returned in the next PR along with tests.

File: src/frontends/tensorflow/src/op/slice.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (rkazants) [reply src/frontends/tensorflow/src/op/slice.cpp:27]: It supports `-1` that we should address.

Comment (itikhono) [reply src/frontends/tensorflow/src/op/slice.cpp:27]: it looks like the only one negative value is supported (-1)
can we simplify this?

**Review by itikhono — CHANGES_REQUESTED** [review]
could you resolve this comment from your prev PR as part of this one?
https://github.com/openvinotoolkit/openvino/pull/12444#discussion_r939903545

File: src/frontends/tensorflow/src/op/slice.cpp (hunk 8 above)
Comment (rkazants) [reply src/frontends/tensorflow/src/op/slice.cpp]: Fixed.

File: src/frontends/tensorflow/src/op/slice.cpp (hunk 7 above)
Comment (rkazants) [reply src/frontends/tensorflow/src/op/slice.cpp]: Simplified. Fixed.

**Comment by rkazants** [comment]
> could you resolve these comments from your prev PR as part of this one? [#12444 (comment)](https://github.com/openvinotoolkit/openvino/pull/12444#discussion_r939903545) [#12444 (comment)](https://github.com/openvinotoolkit/openvino/pull/12444#discussion_r939968658)

Done

File: src/frontends/tensorflow/src/input_model.cpp [hunk 9]
```diff
@@ -170,82 +170,97 @@ std::vector<std::shared_ptr<OpPlace>> InputModel::InputModelTFImpl::get_op_place
...
+                                    producer_name +
+                                    "', expected input port index: " + std::to_string(producer_output_port_idx) + '\n');
+                }
+
+                // skip conditional edges for all operators
+                if (is_conditional_edge(producer_name)) {
+                    continue;
+                }
+
+                // TODO: re-implement the logic below using Place graph structure (with OpPlace, In/OutPortPlace
```
Comment (mvafin) [inline src/frontends/tensorflow/src/input_model.cpp:217]: I think this comment is not valid anymore since we already use `tensor->is_input()` method

**Review by mvafin — APPROVED** [review]

File: src/frontends/tensorflow/src/op/slice.cpp [hunk 10]
```diff
@@ -18,11 +18,24 @@ OutputVector translate_slice_op(const NodeContext& node) {
...
+    // compute stop values in case negative sizes
+    // since TensorFlow supports only -1 among negative sizes
+    // assign stop values to the data shape
+    auto input_shape = make_shared<ShapeOf>(input);
+    auto stop_neg = make_shared<ConvertLike>(input_shape, size);
+
+    // select the correct stop value based on a sign of size value
+    auto zeros = make_shared<Constant>(size.get_element_type(), Shape{}, 0);
+    auto negative_sizes_mask = make_shared<Less>(size, zeros);
+    auto stop = make_shared<Select>(negative_sizes_mask, stop_neg, stop_pos);
```
Comment (itikhono) [inline src/frontends/tensorflow/src/op/slice.cpp:33]: can we add comment that we can try to use FloorMod operation here instead of Select subgraph to improve performance? But additional investigations are needed on benefits of this approach

**Review by itikhono — APPROVED** [review]
LGTM

File: src/frontends/tensorflow/src/input_model.cpp (hunk 9 above)
Comment (rkazants) [reply src/frontends/tensorflow/src/input_model.cpp]: Removed.
